tractor/_spawn.py

```diff
@@ -356,23 +404,53 @@ async def new_proc(
             # awaited and reported upwards to the supervisor.
             await actor_nursery._join_procs.wait()
 
+        finally:
+            # XXX: in the case we were cancelled before the sub-proc
+            # registered itself back we must be sure to try and clean
+            # any process we may have started.
+
+            reaping_cancelled = False
+            cancel_scope = None
+
             if portal in actor_nursery._cancel_after_result_on_exit:
-                cancel_scope = await nursery.start(
-                    cancel_on_completion, portal, subactor, errors)
+                try:
+                    # async with trio.open_nursery() as n:
+                    # n.cancel_scope.shield = True
+                    cancel_scope = await nursery.start(
+                        cancel_on_completion,
+                        portal,
+                        subactor,
+                        errors
+                    )
+                except trio.Cancelled:
+                    # if the reaping task was cancelled we may have hit
+                    # a race where the subproc disconnected before we
+                    # could send it a message to cancel (classic 2 generals)
+                    # in that case, wait shortly then kill the process.
+                    reaping_cancelled = True
+
+                    if proc.is_alive():
+                        with trio.move_on_after(0.1) as cs:
+                            cs.shield = True
+                            await proc_waiter(proc)
+
+                        if cs.cancelled_caught:
+                            proc.terminate()
```

Comment:
I think it is necessary to end this block with an unconditional `raise`. Although most cases would work, I suppose some edge cases would break down the line. For example, consider if you needed to directly cancel the top-level nursery of this function, and then check it's `cancelled_caught` property (similar to the `move_on_after` in this block).  This exception handler would discard the cancel and it would appear to never catch any cancels.

Reply:
Thanks for detailed explanation @richardsheridan, will definitely add!

Reply:
The necessary bit is tagged on the last logic block.